pkg/didcomm/protocol/introduce/service.go

```diff
@@ -26,28 +31,173 @@ const (
 	AckMsgType = IntroduceSpec + "ack"
 )
 
+// message type to store data for eventing. This is retrieved during callback.
+type message struct {
+	Msg           *service.DIDCommMsg
+	NextStateName string
+}
+
+var logger = log.New("aries-framework/introduce/service")
+
 // Service for introduce protocol
 type Service struct {
 	service.Action
 	service.Message
+	store storage.Store
 }
 
 // New returns introduce service
-func New() *Service {
-	return &Service{}
+func New(s storage.Provider) (*Service, error) {
+	store, err := s.OpenStore(Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: storage should be closed e.g s.CloseStore(Introduce)
```

Comment:
I don't think the store should be closed in these lower layers. A store might be shared across multiple packages/layers.

Reply:
`Should be closed` I meant we need to provide a server method `Close` or `Stop`  which will close storage as well (and some other things e.g close listener etc.). We opened it and it should be our responsibility to close it (by providing a server method `Stop`). Or we can replace  `storage.Provider to storage.Store` in that case we do not care about closing it.

Reply:
@troyronda @llorllale @rolsonquadras @fqutishat I would rather use `storage.Store`. Is there any purpose why it was done by giving a whole storage provider?
In that case, it is much easier to satisfy the interface and the client can manipulate with the storage.

Reply:
@soluchok store provider was added as part of https://github.com/hyperledger/aries-framework-go/pull/441 to provide namespace support instead of dumping data in one place.

Reply:
@rolsonquadras  Just tell me why I need to receive a `storage.Provider` instead of a `storage.Store`? What is a use-case for that? I see only complexity

Reply:
It looks like that we provide portability to the service to create a few stores instead of one. There is no protection. I do not think that protocols will communicate with each other using that trick.

Reply:
Protocols should use a Store that is namespaced to the protocol. Having a decoupled design is an objective and keeping storage namespaced is part of that goal.

The thought behind a Storage Provider is that the protocol, itself, should supply the namespace string. Hence the protocol calls OpenStore with that namespace string.

It is up to the Storage Provider to decide how to deal with namespaces.